src/_pytest/subtests.py

```diff
@@ -0,0 +1,387 @@
+        help="Disables subtest output 'dots' in non-verbose mode (EXPERIMENTAL)",
+    )
+    group.addoption(
+        "--no-subtests-reports",
+        action="store_true",
+        dest="no_subtests_reports",
+        default=False,
+        help="Disables subtest output unless it's a failed subtest (EXPERIMENTAL)",
+    )
+
+
+@dataclasses.dataclass
+class SubTestContext:
+    """The values passed to SubTests.test() that are included in the test report."""
+
+    msg: str | None
+    kwargs: dict[str, Any]
+
+
+@dataclasses.dataclass(init=False)
+class SubTestReport(TestReport):
+    context: SubTestContext
+
+    @property
+    def head_line(self) -> str:
+        _, _, domain = self.location
+        return f"{domain} {self.sub_test_description()}"
+
+    def sub_test_description(self) -> str:
+        parts = []
+        if isinstance(self.context.msg, str):
+            parts.append(f"[{self.context.msg}]")
+        if self.context.kwargs:
+            params_desc = ", ".join(
+                f"{k}={v!r}" for (k, v) in sorted(self.context.kwargs.items())
+            )
+            parts.append(f"({params_desc})")
+        return " ".join(parts) or "(<subtest>)"
+
+    def _to_json(self) -> dict[str, Any]:
+        data = super()._to_json()
+        del data["context"]
+        data["_report_type"] = "SubTestReport"
+        data["_subtest.context"] = dataclasses.asdict(self.context)
+        return data
+
+    @classmethod
+    def _from_json(cls, reportdict: dict[str, Any]) -> SubTestReport:
+        report = super()._from_json(reportdict)
+        context_data = reportdict["_subtest.context"]
+        report.context = SubTestContext(
+            msg=context_data["msg"], kwargs=context_data["kwargs"]
+        )
+        return report
+
+    @classmethod
+    def _from_test_report(cls, test_report: TestReport) -> SubTestReport:
+        return super()._from_json(test_report._to_json())
+
+
+@fixture
+def subtests(request: SubRequest) -> Generator[SubTests, None, None]:
+    """Provides subtests functionality."""
+    capmam = request.node.config.pluginmanager.get_plugin("capturemanager")
+    if capmam is not None:
+        suspend_capture_ctx = capmam.global_and_fixture_disabled
+    else:
+        suspend_capture_ctx = nullcontext
+    yield SubTests(request.node.ihook, suspend_capture_ctx, request)
+
+
+# Note: cannot use a dataclass here because Sphinx insists on showing up the __init__ method in the documentation,
+# even if we explicitly use :exclude-members: __init__.
+class SubTests:
+    """Subtests fixture, enables declaring subtests inside test functions via the :meth:`test` method."""
+
+    def __init__(
+        self,
+        ihook: pluggy.HookRelay,
+        suspend_capture_ctx: Callable[[], AbstractContextManager[None]],
+        request: SubRequest,
+    ) -> None:
+        self._ihook = ihook
+        self._suspend_capture_ctx = suspend_capture_ctx
+        self._request = request
+
+    def test(
```

Comment:
Do you believe that `SubTests` will have more user-facing methods in the future? If not, an obvious thing to consider is to replace `test` with `__call__` (and change the fixture name to `subtest`), then the user code is a little nicer, `with subtest(...):` instead of `with subtests.test(...)`.